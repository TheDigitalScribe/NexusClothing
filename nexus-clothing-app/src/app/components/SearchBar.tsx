'use client';

import React from 'react'
import { useState } from 'react';
import { MagnifyingGlassIcon } from '@heroicons/react/16/solid';

export const SearchBar = () => {
  const [query, setQuery] = useState('');
  return (
    <form className="flex items-center">
      <div className="relative">
        <input type="text" value={query} onChange={(e) => setQuery(e.target.value)} placeholder="Search..." className="px-4 py-1 w-40 text-sm md:text-base border border-gray-300 rounded-full focus:outline-blue-400" ></input>
        <div className="absolute inset-y-0 right-0 pr-3 flex items-center pointer-events-none">
          <MagnifyingGlassIcon className="w-5 h-5" />
        </div>
      </div>
    </form>
  )
}
