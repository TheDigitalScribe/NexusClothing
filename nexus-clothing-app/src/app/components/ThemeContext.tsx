'use client'

import React, { createContext, useState, useEffect, useContext, ReactNode } from 'react';

// Union type to hold light/dark values
type Theme = 'light' | 'dark';

interface ThemeContextType {
  theme: Theme;
  toggleTheme: () => void;
}

// ThemeContext for managing the theme in global state
const ThemeContext = createContext<ThemeContextType | undefined>(undefined);

export const ThemeProvider: React.FC<{ children: ReactNode }> = ({ children }) => {
  const [theme, setTheme] = useState<Theme>('light');

  // Retrieves theme from local storage if it is there,
  // Otherwise, set the theme to the browser preference
  useEffect(() => {
    const savedTheme = localStorage.getItem('theme') as Theme;
    if (savedTheme) {
      setTheme(savedTheme);
    } else {
      const systemPreference = window.matchMedia('(prefers-color-scheme: dark)').matches ? 'dark' : 'light';
      setTheme(systemPreference);
    }
  }, []);

  // Apply theme class to body on component mount
  useEffect(() => {
    if (typeof window !== 'undefined') {
      document.body.className = theme === 'light' ? 'light-theme' : 'dark-theme';
    }
  }, [theme]);

  const toggleTheme = () => {
    const newTheme = theme === 'light' ? 'dark' : 'light';
    setTheme(newTheme);
    localStorage.setItem('theme', newTheme);
  };

  return (
    <ThemeContext.Provider value={{ theme, toggleTheme }}>
      {children}
    </ThemeContext.Provider>
  );
};

// Exports custom hook for accessing theme context
export const useTheme = () => {
  const context = useContext(ThemeContext);
  if (!context) {
    throw new Error('useTheme must be used within a ThemeProvider');
  }
  return context;
};

export default ThemeContext;