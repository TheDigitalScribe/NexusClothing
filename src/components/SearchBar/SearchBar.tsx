"use client";

import React from 'react'
import { useState } from 'react';
import { IoSearch } from "react-icons/io5";

export const SearchBar: React.FC = () => {
  const [query, setQuery] = useState<string>("");

  const handleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
  };

  return (
    <form className="flex items-center" role="search" onSubmit={handleSubmit}>
      <div className="relative">
        <input type="text" value={query} onChange={(e) => setQuery(e.target.value)} aria-label="Search" placeholder="Search..." className="px-4 py-1 w-40 text-sm md:text-base border border-gray-400 dark:border-gray-700 rounded-full bg-gray-200 dark:bg-gray-900"></input>
        <div className="absolute inset-y-0 right-0 pr-3 flex items-center pointer-events-none">
          <IoSearch className="w-5 h-5" />
        </div>
      </div>
    </form>
  );
};
