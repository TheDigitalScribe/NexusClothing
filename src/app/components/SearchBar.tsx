'use client';

import React from 'react'
import { useState } from 'react';
import { useTheme } from './ThemeContext';
import { ThemeContextType } from '../types';
import { MagnifyingGlassIcon } from '@heroicons/react/16/solid';

export const SearchBar: React.FC = () => {
  const [query, setQuery] = useState<string>('');
  const { theme } = useTheme() as ThemeContextType;

  return (
    <form className="flex items-center">
      <div className="relative">
        <input type="text" value={query} onChange={(e) => setQuery(e.target.value)} placeholder="Search..." className={`px-4 py-1 w-40 text-sm md:text-base border rounded-full focus:outline-blue-400 searchbar-transition ${theme == "dark" ? "dark-theme bg-gray-300 input-dark" : "light-theme bg-gray-700 border-gray-700 input-light"}`}></input>
        <div className="absolute inset-y-0 right-0 pr-3 flex items-center pointer-events-none">
          <MagnifyingGlassIcon className="w-5 h-5" />
        </div>
      </div>
    </form>
  )
}
