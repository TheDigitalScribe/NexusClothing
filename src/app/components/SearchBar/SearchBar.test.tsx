// SearchBar.test.tsx
import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { SearchBar } from './SearchBar';
import ThemeContext from '../ThemeContext';

// Mock the HeroIcons component
jest.mock('@heroicons/react/16/solid', () => ({
  MagnifyingGlassIcon: () => <div data-testid="search-icon">SearchIcon</div>,
}));

describe('SearchBar', () => {
  const renderWithTheme = (theme: 'light' | 'dark') => {
    return render(
      <ThemeContext.Provider value={{ theme, toggleTheme: jest.fn() }}>
        <SearchBar />
      </ThemeContext.Provider>
    );
  };

  describe('Rendering', () => {
    it('renders correctly in light theme', () => {
      const { container } = renderWithTheme('light');
      expect(container).toMatchSnapshot();
    });

    it('matches snapshot', () => {
      const { container } = render(<SearchBar />);
      expect(container).toMatchSnapshot();
    });

    it('renders correctly in dark theme', () => {
      const { container } = renderWithTheme('dark');
      expect(container).toMatchSnapshot();
    });

    it('renders search input with correct placeholder', () => {
      renderWithTheme('light');
      expect(screen.getByPlaceholderText('Search...')).toBeInTheDocument();
    });

    it('renders search icon', () => {
      renderWithTheme('light');
      expect(screen.getByTestId('search-icon')).toBeInTheDocument();
    });
  });

  describe('Theming', () => {
    it('applies correct classes in light theme', () => {
      renderWithTheme('light');
      const input = screen.getByRole('textbox', { name: 'Search' });
      expect(input).toHaveClass('light-theme', 'bg-gray-700', 'border-gray-700', 'input-light');
    });

    it('applies correct classes in dark theme', () => {
      renderWithTheme('dark');
      const input = screen.getByRole('textbox', { name: 'Search' });
      expect(input).toHaveClass('dark-theme', 'bg-gray-300', 'input-dark');
    });
  });

  describe('Input Functionality', () => {
    it('updates input value on user typing', async () => {
      const user = userEvent.setup();
      renderWithTheme('light');
      const input = screen.getByPlaceholderText('Search...') as HTMLInputElement;

      await user.type(input, 'test search');
      expect(input.value).toBe('test search');
    });

    it('handles empty input correctly', async () => {
      const user = userEvent.setup();
      renderWithTheme('light');
      const input = screen.getByPlaceholderText('Search...') as HTMLInputElement;

      await user.type(input, 'test');
      await user.clear(input);
      expect(input.value).toBe('');
    });
  });

  describe('Form Behavior', () => {
    it('prevents default form submission', () => {
      renderWithTheme('light');
      const form = screen.getByRole('search');

      const handleSubmit = jest.fn();
      form.addEventListener('submit', handleSubmit);

      fireEvent.submit(form);

      expect(handleSubmit).toHaveBeenCalledTimes(1);
    });

    it('has correct form attributes and classes', () => {
      renderWithTheme('light');
      const form = screen.getByRole('search');

      expect(form).toHaveClass('flex', 'items-center');

      expect(form).toHaveAttribute('role', 'search');
    });

    it('maintains form state after submission', async () => {
      const user = userEvent.setup();
      renderWithTheme('light');

      const input = screen.getByRole('textbox', { name: 'Search' });
      const form = screen.getByRole('search');

      await user.type(input, 'test search');

      fireEvent.submit(form);

      expect(input).toHaveValue('test search');
    });

    it('handles form submission with enter key', async () => {
      const user = userEvent.setup();
      renderWithTheme('light');

      const mockPreventDefault = jest.fn();
      const form = screen.getByRole('search');
      const input = screen.getByRole('textbox', { name: 'Search' });

      form.addEventListener('submit', (e) => {
        e.preventDefault = mockPreventDefault;
      });

      await user.type(input, 'test search{enter}');

      expect(mockPreventDefault).toHaveBeenCalled();
    });
  });

  describe('Accessibility', () => {
    it('input is keyboard accessible', async () => {
      const user = userEvent.setup();
      renderWithTheme('light');

      await user.tab();
      expect(screen.getByPlaceholderText('Search...')).toHaveFocus();
    });

    it('has proper focus styles', () => {
      renderWithTheme('light');
      const input = screen.getByPlaceholderText('Search...');
      expect(input).toHaveClass('focus:outline-blue-400');
    });
  });

  describe('Responsive Design', () => {
    it('has correct base and responsive text classes', () => {
      renderWithTheme('light');
      const input = screen.getByPlaceholderText('Search...');
      expect(input).toHaveClass('text-sm', 'md:text-base');
    });

    it('has correct width class', () => {
      renderWithTheme('light');
      const input = screen.getByPlaceholderText('Search...');
      expect(input).toHaveClass('w-40');
    });
  });

  describe('State Management', () => {
    it('maintains state between renders', async () => {
      const user = userEvent.setup();
      const { rerender } = renderWithTheme('light');
      const input = screen.getByPlaceholderText('Search...') as HTMLInputElement;

      await user.type(input, 'test');
      expect(input.value).toBe('test');

      rerender(
        <ThemeContext.Provider value={{ theme: 'light', toggleTheme: jest.fn() }}>
          <SearchBar />
        </ThemeContext.Provider>
      );

      expect(input.value).toBe('test');
    });
  });
});
