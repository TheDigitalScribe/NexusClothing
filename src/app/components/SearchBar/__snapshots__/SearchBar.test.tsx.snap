// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar Rendering renders correctly in dark theme 1`] = `
<div>
  <form
    class="flex items-center"
    role="search"
  >
    <div
      class="relative"
    >
      <input
        aria-label="Search"
        class="px-4 py-1 w-40 text-sm md:text-base border rounded-full focus:outline-blue-400 searchbar-transition dark-theme bg-gray-300 input-dark"
        placeholder="Search..."
        type="text"
        value=""
      />
      <div
        class="absolute inset-y-0 right-0 pr-3 flex items-center pointer-events-none"
      >
        <div
          data-testid="search-icon"
        >
          SearchIcon
        </div>
      </div>
    </div>
  </form>
</div>
`;

exports[`SearchBar Rendering renders correctly in light theme 1`] = `
<div>
  <form
    class="flex items-center"
    role="search"
  >
    <div
      class="relative"
    >
      <input
        aria-label="Search"
        class="px-4 py-1 w-40 text-sm md:text-base border rounded-full focus:outline-blue-400 searchbar-transition light-theme bg-gray-700 border-gray-700 input-light"
        placeholder="Search..."
        type="text"
        value=""
      />
      <div
        class="absolute inset-y-0 right-0 pr-3 flex items-center pointer-events-none"
      >
        <div
          data-testid="search-icon"
        >
          SearchIcon
        </div>
      </div>
    </div>
  </form>
</div>
`;
